### PR TITLE
add actionButton in frontend

### LINKED ISSUE #1439: Add universal action in frontend
Add a universal button that can create a modal, when the buttons are pressed, the state can change and a backend request can occur

### PR DESCRIPTION
**Pull request by Arrgentum**
## Whats added
- Add actionButton class 
- Add modals
- Integrated actionButton for deleting organizations/projects
## Issue(#1439)
This pull request close #1439

### CODE REVIEW

**Review by nulls — APPROVED** [review]

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/organizations/OrganizationSettingsMenu.kt [hunk 1]
```diff
@@ -139,17 +159,36 @@ private fun organizationSettingsMenu() = FC<OrganizationSettingsMenuProps> { pro
...
-                    deleteOrganizationButton {
-                        organizationName = props.organizationName
-                        onDeletionSuccess = {
-                            window.location.href = "${window.location.origin}/"
-                        }
+                    actionButton {
+                        title = "WARNING: Delete Organization"
+                        errorTitle = "You cannot delete a ${props.organizationName}"
+                        message = "Are you sure you want to delete the organization ${props.organizationName}?"
+                        clickMessage = "Change to ban mode"
```
[the remark this answers is not in the document]
Comment (sanyavertolet) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/organizations/OrganizationSettingsMenu.kt:166]: To my mind checkbox is much prettier, but to my mind the text should be like `Also ban this organization`

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/organizations/OrganizationSettingsMenu.kt [hunk 2]
```diff
@@ -139,17 +159,36 @@ private fun organizationSettingsMenu() = FC<OrganizationSettingsMenuProps> { pro
...
+                                    window.closeWindow()
+                                }
+                                buttonBuilder("Cancel") {
+                                    window.closeWindow()
+                                }
+                            }
+                        }
+                        onActionSuccess = { _ ->
+                            window.location.href = "${window.location.origin}/"
+                        }
```
Comment (sanyavertolet) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/organizations/OrganizationSettingsMenu.kt:186]: Do we still use `window.location`?

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/projects/ProjectSettingsMenu.kt [hunk 3]
```diff
@@ -267,14 +257,36 @@ private fun projectSettingsMenu() = FC<ProjectSettingsMenuProps> { props ->
...
+                            title = "WARNING: You want to delete a project"
+                            errorTitle = "You cannot delete ${props.project.name}"
+                            message = "Are you sure you want to delete a $projectPath?"
+                            clickMessage = "Change to ban mode"
+                            onActionSuccess = { _ ->
+                                window.location.href = "${window.location.origin}/"
+                            }
+                            buttonStyleBuilder = { childrenBuilder ->
+                                with(childrenBuilder) {
+                                    +"Delete ${props.project.name}"
```
Comment (sanyavertolet) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/projects/ProjectSettingsMenu.kt:270]: I think messages could be better but it is not that important

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/modal/ModalBuilder.kt [hunk 4]
```diff
@@ -71,6 +72,34 @@ fun ChildrenBuilder.displayModal(
...
+ * @param isOpen modal openness indicator - should be in state
+ * @param modalStyle that will be applied to react modal
+ * @param onCloseButtonPressed callback that will be applied to `X` button in the top-right corner
+ * @param buttonBuilder lambda that generates several buttons, must contain either [button] or [buttonBuilder]
+ * @param title of the modal that will be shown in top-left corner
+ * @param message main text that will be shown in the center of modal
+ * @param clickBuilder lambda that generates several click in modal
+ */
+@Suppress("LongParameterList", "TOO_MANY_PARAMETERS")
+fun ChildrenBuilder.displayModalWithClick(
```
Comment (sanyavertolet) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/modal/ModalBuilder.kt:87]: With click?

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/utils/Action.kt [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+        }
+    )
+}
+
+/**
+ * Button with modal for some action
+ *
+ * @return noting
+ */
+external interface ActionProps : Props {
```
Comment (sanyavertolet) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/utils/Action.kt:99]: ActionProps is not clear - ButtonWithActionProps should be better imo

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/utils/Action.kt [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+                                value = isClickMode
+                                id = "click"
+                                checked = isClickMode
+                                onChange = {
+                                    setClickMode(!isClickMode)
+                                }
+                            }
+                        }
+                        div {
+                            ReactHTML.label {
```
Comment (sanyavertolet) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/utils/Action.kt:81]:
```suggestion
                            label {
```

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/utils/Action.kt [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+            }
+        },
+        clickBuilder = {
+            if (props.conditionClick && !isError) {
+                div {
+                    className = ClassName("d-sm-flex justify-content-center form-check")
+                    div {
+                        className = ClassName("d-sm-flex justify-content-center form-check")
+                        div {
+                            ReactHTML.input {
```
Comment (sanyavertolet) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/utils/Action.kt:69]:
```suggestion
                            input {
```

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/utils/Action.kt [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+                                className = ClassName("click")
```
Comment (sanyavertolet) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/utils/Action.kt:70]: Is there such class in bootstrap?

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/modal/ModalBuilder.kt [hunk 9]
```diff
@@ -228,3 +257,98 @@ fun ChildrenBuilder.modalBuilder(
...
+                }
+            }
+            className = ClassName("card card-body mt-0 p-0")
+            div {
+                className = ClassName("modal-body")
+                bodyBuilder()
+            }
+            div {
+                className = ClassName("d-sm-flex justify-content-center form-check")
+                clickBuilder()
```
Comment (sanyavertolet) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/modal/ModalBuilder.kt:346]: Shouldn't it be inside `modal-body`?

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/modal/ModalBuilder.kt [hunk 10]
```diff
@@ -228,3 +257,98 @@ fun ChildrenBuilder.modalBuilder(
...
+ *
+ * @param onCloseButtonPressed callback that will be applied to `X` button in the top-right corner
+ * @param buttonBuilder lambda that generates several buttons, must contain either [button] or [buttonBuilder]
+ * @param clickBuilder lambda that generates several click in modal
+ * @param title of the modal that will be shown in top-left corner
+ * @param message main text that will be shown in the center of modal
+ * @param clickBuilder
+ */
+@Suppress("TOO_MANY_PARAMETERS", "LongParameterList")
+fun ChildrenBuilder.modalBuilderWithClick(
```
Comment (sanyavertolet) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/modal/ModalBuilder.kt:272]: Do you really need this method? I mean it seems you use this one only once as well as the one with three builders (the one below).

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/organizations/OrganizationSettingsMenu.kt [hunk 11]
```diff
@@ -139,17 +159,36 @@ private fun organizationSettingsMenu() = FC<OrganizationSettingsMenuProps> { pro
...
                 div {
                     className = ClassName("d-sm-flex align-items-center justify-content-center p-3")
-                    deleteOrganizationButton {
-                        organizationName = props.organizationName
-                        onDeletionSuccess = {
-                            window.location.href = "${window.location.origin}/"
-                        }
+                    actionButton {
+                        title = "WARNING: Delete Organization"
+                        errorTitle = "You cannot delete a ${props.organizationName}"
```
Comment (0x6675636b796f75676974687562) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/organizations/OrganizationSettingsMenu.kt:164]: Please don't use _a_/_an_ with _proper names_. "An Andrey" or "an Alexey" doesn't sound right.

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/projects/ProjectSettingsMenu.kt [hunk 12]
```diff
@@ -267,14 +257,36 @@ private fun projectSettingsMenu() = FC<ProjectSettingsMenuProps> { props ->
...
                     div {
                         className = ClassName("col-3 d-sm-flex align-items-center justify-content-center")
-                        button {
-                            type = ButtonType.button
-                            className = ClassName("btn btn-sm btn-danger")
-                            disabled = !props.selfRole.hasDeletePermission()
-                            onClick = {
-                                deletionModalOpener.openWindow()
+                        actionButton {
+                            title = "WARNING: You want to delete a project"
```
Comment (0x6675636b796f75676974687562) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/projects/ProjectSettingsMenu.kt:261]:
```suggestion
                            title = "WARNING: About delete a project..."
```

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/projects/ProjectSettingsMenu.kt [hunk 13]
```diff
@@ -267,14 +257,36 @@ private fun projectSettingsMenu() = FC<ProjectSettingsMenuProps> { props ->
...
+                            errorTitle = "You cannot delete ${props.project.name}"
+                            message = "Are you sure you want to delete a $projectPath?"
```
Comment (0x6675636b796f75676974687562) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/projects/ProjectSettingsMenu.kt:263]: _A/an_ with _proper names_.

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/OrganizationAdminView.kt [hunk 14]
```diff
@@ -66,24 +62,41 @@ internal class OrganizationAdminView : AbstractView<Props, OrganizationAdminStat
...
-                                id = "delete-organization-$organizationName"
-                                classes = deleteButtonClasses
-                                tooltipText = "Delete the organization"
-                                elementChildren = { childrenBuilder ->
+                            val organization = cellProps.value
+                            val organizationName = organization.name
+
+                            actionButton {
+                                title = "WARNING: Delete Organization"
+                                errorTitle = "You cannot delete a $organizationName"
```
Comment (0x6675636b796f75676974687562) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/OrganizationAdminView.kt:70]: _A/an_ with _proper names_.

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/usersettings/UserSettingsOrganizationsMenuView.kt [hunk 15]
```diff
@@ -56,17 +64,34 @@ class UserSettingsOrganizationsMenuView : UserSettingsView() {
...
                                 val role = state.userInfo?.name?.let { organizationDto.userRoles[it] } ?: Role.NONE
                                 if (role.isHigherOrEqualThan(Role.OWNER)) {
-                                    deleteOrganizationButton {
-                                        organizationName = organizationDto.name
-                                        onDeletionSuccess = {
-                                            setState { selfOrganizationDtos = selfOrganizationDtos.minusElement(organizationDto) }
-                                        }
+                                    actionButton {
+                                        title = "WARNING: Delete Organization"
+                                        errorTitle = "You cannot delete a ${organizationDto.name}"
```
Comment (0x6675636b796f75676974687562) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/usersettings/UserSettingsOrganizationsMenuView.kt:69]: _A/an_ with _proper names_.

Comment (Arrgentum) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/usersettings/UserSettingsOrganizationsMenuView.kt]: Doen

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/OrganizationView.kt [hunk 16]
```diff
@@ -206,24 +206,41 @@ class OrganizationView : AbstractView<OrganizationProps, OrganizationViewState>(
...
-                                    id = "delete-project-$projectName"
-                                    classes = deleteButtonClasses
-                                    tooltipText = "Delete the project"
-                                    elementChildren = { childrenBuilder ->
+                                val project = cellProps.row.original
+                                val projectName = project.name
+
+                                actionButton {
+                                    title = "WARNING: Delete Project"
+                                    errorTitle = "You cannot delete a $projectName"
```
[the remark this answers is not in the document]
Comment (Arrgentum) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/OrganizationView.kt]: Done

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/OrganizationAdminView.kt (hunk 14 above)
Comment (Arrgentum) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/OrganizationAdminView.kt]: Done

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/projects/ProjectSettingsMenu.kt (hunk 13 above)
Comment (Arrgentum) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/projects/ProjectSettingsMenu.kt]: Done

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/projects/ProjectSettingsMenu.kt (hunk 12 above)
Comment (Arrgentum) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/projects/ProjectSettingsMenu.kt]: Done

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/organizations/OrganizationSettingsMenu.kt (hunk 11 above)
Comment (Arrgentum) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/organizations/OrganizationSettingsMenu.kt]: Done

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/modal/ModalBuilder.kt (hunk 10 above)
Comment (Arrgentum) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/modal/ModalBuilder.kt]: I deleted 1 method
Now there are 2 of them left

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/utils/Action.kt (hunk 8 above)
Comment (Arrgentum) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/utils/Action.kt]: Renamed

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/utils/Action.kt (hunk 7 above)
Comment (Arrgentum) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/utils/Action.kt]: Done

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/utils/Action.kt (hunk 6 above)
Comment (Arrgentum) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/utils/Action.kt]: Done

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/utils/Action.kt (hunk 5 above)
Comment (Arrgentum) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/utils/Action.kt]: Done

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/modal/ModalBuilder.kt (hunk 4 above)
Comment (Arrgentum) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/modal/ModalBuilder.kt]: Renamed to ButtonWithActionProps

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/organizations/OrganizationSettingsMenu.kt (hunk 2 above)
Comment (Arrgentum) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/organizations/OrganizationSettingsMenu.kt]: redone with navigate

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/organizations/OrganizationSettingsMenu.kt (hunk 11 above)
[the remark this answers is not in the document]
Comment (Arrgentum) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/organizations/OrganizationSettingsMenu.kt]: Yes.
In line 29 in ```Action.kt``` :
``` kotlin
      setDisplayTitle(props.errorTitle)
      setDisplayMessage(response.unpackMessage())
      setError(true)
      windowOpenness.openWindow()
```

After an unsuccessful request, we put an ```errorMessage``` in the message field. We pass the title of this message to ```errorTitle```

File: save-agent/logs.txt [hunk 17]
```diff
@@ -0,0 +1 @@
+0
```
[the remark this answers is not in the document]
Comment (Arrgentum) [reply save-agent/logs.txt:1]: Most likely diktatFix did it

[the remark this answers is not in the document]
Comment (Arrgentum) [reply save-agent/logs.txt]: Deleted

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/modal/ModalBuilder.kt (hunk 9 above)
Comment (Arrgentum) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/modal/ModalBuilder.kt]: Done

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/modal/ModalBuilder.kt (hunk 10 above)
Comment (Arrgentum) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/modal/ModalBuilder.kt]: Now there is 1 added modal left

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/organizations/OrganizationSettingsMenu.kt [hunk 18]
```diff
@@ -67,6 +72,23 @@ external interface OrganizationSettingsMenuProps : Props {
...
 }
 
+/**
+ * Makes a call to change project status
+ *
+ * @param organizationName name of the organization whose status will be changed
+ * @param status is new status
+ * @return response
+ */
+fun responseChangeOrganizationStatus(status: OrganizationStatus, organizationName: String): suspend WithRequestStatusContext.() -> Response = {
```
Comment (0x6675636b796f75676974687562) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/organizations/OrganizationSettingsMenu.kt:82]: I'd change the order of parameters to match that in the API URL:

```suggestion
fun responseChangeOrganizationStatus(organizationName: String, status: OrganizationStatus): suspend WithRequestStatusContext.() -> Response = {
```

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/organizations/OrganizationSettingsMenu.kt [hunk 19]
```diff
@@ -67,6 +72,23 @@ external interface OrganizationSettingsMenuProps : Props {
     var onCanCreateContestsChange: (Boolean) -> Unit
 }
 
+/**
+ * Makes a call to change project status
+ *
+ * @param organizationName name of the organization whose status will be changed
+ * @param status is new status
+ * @return response
```
Comment (0x6675636b796f75676974687562) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/organizations/OrganizationSettingsMenu.kt:80]: Please document the return type here.

```kotlin
suspend WithRequestStatusContext.() -> Response
```

is not a response &mdash; at least, you should say this is a _lazy_ response, if I understand your intention correctly.

**Review by 0x6675636b796f75676974687562 — APPROVED** [review]

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/organizations/OrganizationSettingsMenu.kt (hunk 19 above)
Comment (Arrgentum) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/organizations/OrganizationSettingsMenu.kt]: Done

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/organizations/OrganizationSettingsMenu.kt (hunk 18 above)
Comment (Arrgentum) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/organizations/OrganizationSettingsMenu.kt]: Done

**Pull request merged**